### ISSUE TITLE
Improving rewards and punishments with checkpoint macro blocks

### ISSUE THREAD

**Issue by brunoffranca** [nimiq/core-rs-albatross#183]
# Problem
On the current implementation we have several undesirable issues with the rewards and punishments:
- The rewards are only distributed a full epoch after the current epoch ends. This means that when you get selected to be a validator you will only get your reward two full epochs later. We expect epochs to be ~6h, this results in waiting 12h.
- Validators that forked don't get barred from producing micro blocks. This was decided since on a fork, different fork proofs could get submitted, resulting on a situation where the two forks have different slashed sets.
- Punishment for a view change is the loss of the entire epoch's reward and getting barred from producing blocks for the rest of the epoch. But there's no difference to the network between a validator that is offline (or purposely not producing blocks) and a validator that is getting DDoS'ed. If someone wants to hurt a validator (i.e. a competing pool) DDoSing them for 5m would be enough to make them lose the reward of an entire epoch. It is a very cheap and effective attack.

# Solution
Now that we have checkpoint macro blocks, we can improve on all these points.
- Distribute rewards every batch instead of every epoch. The rewards are distributed at the end of the next batch (just like we did with epochs), but now this happens every macro block (checkpoint and election) instead of happening only at election macro blocks. Reward slashes now also happen every macro block, obviously. [This solves issue 1]
- Bar micro block production for both view changes and forks. But now they only happen after the next macro block. So, if a view change or fork proof happens during a batch, at the end of that batch (at the next macro block) the misbehaving validator will get "parked" (currently park/unparking is already implemented on the staking contract but with a slightly different functionality). While he is parked, he cannot produce micro blocks or propose macro blocks (but can still participate in the macro block production). [This solves issue 2]
- We follow the current specification for the park/unpark mechanism where the validator gets "potentially parked" right at the next macro block (can't produce micro blocks, can still get selected as validator) and get "effectively parked" at the macro block after the next one (can't produce micro blocks, can't get selected as validator).
- A validator that has been parked can send a transaction to unpark himself that will take effect at the next macro block. Just like the current park/unpark mechanism a validator can send an unstaking transaction during the batch when he misbehaved (before he gets "potentially parked") and in that case will not get "parked" in the next macro block (technically he gets parked and unparked in the same block). Now an attacker needs to be constantly DDoSing a validator in order to have him lose the reward of an entire epoch. [This solves issue 3]

# Summary

1. Have rewards and punishments apply at the end of every batch.
2. Apply the park/unpark mechanism to every batch.
3. Reintroduce micro block barring for forkers.

### Nomenclature
Macro blocks have two types: checkpoint and election. Both are produced with Tendermint. Only election blocks change the validator list.
A batch consists of several micro blocks and the macro block (checkpoint or election) that closes it. Or the time between two consecutive macro blocks.
An epoch consists of several batches where only the last batch ends with an election macro block. Or the time between two consecutive election macro blocks.
A diagram of the above situation: 
 m -> m-> m -> CM -> m -> m-> m -> CM -> m -> m-> m -> EM

**Comment by maestroi** [comment]
This is really good Thank you!

**Comment by paberr** [comment]
Here's the result of today's discussion and how we intend to work with misbehaving validators in the future:

# Slashing Behaviour

## General Behaviour:
- On misbehaviour (view changes, fork proofs) validators are marked as *disabled slot*, *lost reward* and *parked* in the staking contract. The first two sets apply only to slots, while *parking* applies to the whole validator. 
- Except for this, the misbehaviour doesn’t have any implications on the current batch.
- At the transition between batches (i.e., in the macro blocks), the current set of disabled slots is retrieved from the staking contract and from then on has impact on which slots are eligible for block production. That means that both fork proofs and view changes bar validators from producing blocks in a certain slot starting from the following batch.
- Validators can send an unslashing transaction at any time, which removes them from both the *parked* and the *disabled slots* sets. If the transaction is sent between the misbehaviour and the next macro block, no barring from block production occurs.
- The misbehaving slots will never receive rewards for the batch it occurred in. More precisely, all slots in *lost reward* bitwise or *disabled slots* will not receive rewards. Note that these rewards of batch i are payed out at the end of batch i+1 only. After paying out the rewards for a batch, its *lost reward* set is reset. We thus need to keep track of two sets for lost rewards (current batch and previous batch).
- Assuming misbehaviour in epoch j, if no unslashing transaction happens within epoch j or j+1, the whole validator will be put inactive at the end of epoch j+1. This is what the *parking* set is for.

## Nomenclature:
- Disabled slots: The set of slots that is punished for misbehaviour by barring these slots from block production. This set is only reset after each election block.
- Lost reward: The set of slots that tracks which slots are definitely ineligible to receive rewards in the next batch. Note that every slot on the *disabled slots* set is also not receiving any rewards. The *lost rewards* set is reset after distributing the respective rewards. Unslashing transactions do not have an impact on this set. This ensures that misbehaving slots lose their rewards at least once.
- Unslashing transaction: Also called unparking transaction. This is a transaction a validator can send to prevent them from being put inactive and reactivates their ability to produce blocks (i.e., removes them from the *parked* set and *disabled slots* set).
- Parking: Parking denotes the temporary state of a validator being marked to be inactivated. If a validator is parked, it will be put inactive after one full epoch of grace period. A validator can prevent this by sending an unparking transaction.

## Details:
- Slashes are stored in the staking contract (applied with view change inherent) and can be unslashed there
- The staking contract needs to store (slot, public key) in order to be able to remove slots from the *disabled slots* set upon unslashing transactions (which only contain the public key).
- We need to make sure you always loose rewards even if you remove yourself from the slashed set for block production, that’s why we distinguish between *lost rewards* and *disabled slots*.
- The *disabled slots* set (as well as the *lost reward* and *parked* sets) is updated immediately with the inherent generated by the misbehaviour. The effective set used for the slot calculation, however, is only read after each macro block. Thus, while the staking contract immediately reflects the changes, the punishment only comes into effect after the end of the batch.
- This delay keeps the effective *disabled slots* set constant during each batch and allows to predict the correct slot owner at any point during the batch when the previous seed is known.
- For optimisation, we put the *disabled slots* set of the closing batch into the inherents of the macro block.

## Example:
Epoch j, batch i, block k: A view change occurs.
This view change generates an inherent that modifies the staking contract, adding the slot to the lost reward and disabled slots sets and putting the validator on the parked list.

At the next macro block:
The effective disabled slots for block production are read from the staking contract. These will be active from the beginning of the next batch (barring the slot from producing blocks).
Unslashing in batch i would result in the slot not being part of the effective disabled slots for block production read at this macro block.
Unslashing in batch i+1 would take effect only after next macro block (in batch i+2).

Parking works independently and follows the same pattern it does now.
Without an unslashing transaction, the validator will be put inactive at the election block of epoch j+1 (taking effect from epoch j+2).

**Note: We want that the validator cannot be chosen for epoch j+2, we need to make sure that the validator selection is run after committing the accounts tree state.**

**Comment by brunoffranca** [comment]
I think you need to calculate the rewards based on the disabled set for the _previous_ batch, like you do for the lost rewards set.
Otherwise, if at batch x I'm in the disabled set but not in the previous lost rewards set and unslash myself, then I'll get the rewards for that batch without even being allowed to produce blocks.
We already keep track of the previous disabled set anyway, we just need to make sure that we calculate the rewards based on it.
